core/services/relay/evm/chain_reader.go

```diff
@@ -0,0 +1,197 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	relaytypes "github.com/smartcontractkit/chainlink-relay/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+// constructor for ChainReader, returns nil if there is any error
+func newChainReader(lggr logger.Logger, chain evm.Chain, ropts *types.RelayOpts) (*chainReader, error) {
+	relayConfig, err := ropts.RelayConfig()
+	if err != nil {
+		return nil, fmt.Errorf("Failed parsing RelayConfig: %w", err)
+	}
+
+	if relayConfig.ChainReader == nil {
+		return nil, relaytypes.ErrorChainReaderUnsupported{}
+	}
+
+	if err = ValidateChainReaderConfig(*relayConfig.ChainReader); err != nil {
+		return nil, fmt.Errorf("invalid ChainReader configuration: %w", err)
+	}
+
+	return NewChainReaderService(lggr, chain.LogPoller())
+}
+
+func ValidateChainReaderConfig(cfg types.ChainReaderConfig) error {
```

Comment:
Does this need to be exported?
```suggestion
func validateChainReaderConfig(cfg types.ChainReaderConfig) error {
```